### PR TITLE
A more complete BLAST cmd API

### PR DESCRIPTION
**Pull request by eparejatobes**

### CODE REVIEW

**Comment by eparejatobes** [comment]
Mostly done. Still missing
- [ ] most command-specific options are still missing (protein-related stuff for blastp, translation for tblastn, etc)
- [ ] complete set of output fields, together with accurate types for which commands (blastn, blastp, ...) support them
- [x] nicer types and syntax for building full commands (blastn args options)
- [x] differentiate between arguments and options (require input files etc)

**Comment by eparejatobes** [comment]
ping @marina-manrique @laughedelic review if possible. @marina-manrique maybe you can do most of what's missing.

**Comment by marina-manrique** [comment]
:+1: maybe we can discuss later how I can go ahead with this

**Comment by laughedelic** [comment]
I'll take a look at the code later today

File: src/main/scala/blast.scala [hunk 1]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+  case object tblastx       extends BlastCommand("blastp");   type tblastx  = tblastx.type
+  case object makeblastdb   extends BlastCommand("makeblastdb"); type makeblastdb = makeblastdb.type
+
+  type AllBlasts  = blastn :~: blastp :~: blastx :~: tblastn :~: tblastx :~: ∅
+  val allBlasts   = blastn :~: blastp :~: blastx :~: tblastn :~: tblastx :~: ∅
+
+  sealed trait AnyBlastOption {
+
+    type Commands <: AnyTypeSet.Of[AnyBlastCommand]
+    // val commands: Commands
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: I guess, this value should be uncommented

File: src/main/scala/blast.scala [hunk 2]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+  case class db(val file: File)             extends BlastOption(
+    allBlasts,
+    Seq("-db", file.getCanonicalPath().toString)
+  )
+  case class query(val file: File)          extends BlastOption(
+    allBlasts,
+    Seq("-query", file.getCanonicalPath().toString)
+  )
+  case class out(val file: File)            extends BlastOption(
+    makeblastdb :~: allBlasts,
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: So `makeblastdb <: BlastCommand` is not a blast command?

File: src/main/scala/blast.scala [hunk 3]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+  )
+
+  case object task {
+
+    // TODO finish writing all tasks here
+    case object megablast                 extends BlastTask( blastAPI.blastn, Seq("-task", "megablast") )
+    case object dcMegablast               extends BlastTask( blastAPI.blastn, Seq("-task", "dc-megablast") )
+    case object blastn                    extends BlastTask( blastAPI.blastn, Seq("-task", "blastn") )
+    case object blastnShort               extends BlastTask( blastAPI.blastn, Seq("-task", "blastn-short") )
+    case object rmblastn                  extends BlastTask( blastAPI.blastn, Seq("-task", "rmblastn") )
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: if all tasks have this `"-task"` in the beginning, better to move it to the constructor

File: src/main/scala/blast.scala [hunk 4]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+    case object tblastnFast
+  }
+
+  sealed trait BlastDBInputType
+  object dbInputType {
+
+    case object asn1_bin  extends BlastDBInputType
+    case object asn1_txt  extends BlastDBInputType
+    case object blastdb   extends BlastDBInputType
+    case object fasta     extends BlastDBInputType
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: Could be properties with `Raw = File`, then passing it to a command would be more convenient

File: src/main/scala/blast.scala (hunk 2 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: ?

File: src/main/scala/blast.scala [hunk 5]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+    trait ForCommands[Cmmnds <: AnyTypeSet.Of[AnyBlastCommand]] extends AnyOutputField { type Commands = Cmmnds }
+
+    // means Query Seq-id
+    case object qseqid    extends OutputField[String] with ForCommands[AllBlasts]         { val label = toString }
+    // means Query GI
+    case object qgi       extends OutputField[String] with ForCommands[AllBlasts]         { val label = toString }
+    // means Query accesion
+    case object qacc      extends OutputField[String] with ForCommands[AllBlasts]         { val label = toString }
+    // means Query accesion.version
+    case object qaccver   extends OutputField[Int] with ForCommands[AllBlasts]            { val label = toString }
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: if they all have `val label = toString` put it in the constructor (with `lazy`)

File: src/main/scala/blast.scala [hunk 6]
```diff
@@ -62,28 +62,293 @@ abstract class Blast(val version: String) extends Bundle { blast =>
...
+  }
+
+  import ohnosequences.cosas.ops.typeSets.CheckForAll
+
+  def validOptions[
+    C <: AnyBlastCommand,
+    Opts <: AnyTypeSet.Of[AnyBlastOption]
+  ](cmd: C, opts: Opts)(implicit ev: CheckForAll[Opts, OptionFor[C]]): Opts = {
+
+    opts.checkForAll[OptionFor[C]]; opts
```
Comment (laughedelic) [inline src/main/scala/blast.scala]: `opts.checkForAll` — what's this for? you already checked it by `ev` (actually that method returns just the same as `ev`)

File: src/main/scala/blast.scala (hunk 2 above)
Comment (laughedelic) [inline src/main/scala/blast.scala]: why isn't it in `allBlasts`?

File: src/main/scala/blast.scala (hunk 5 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: I always forget which of those things work and which not.

File: src/main/scala/blast.scala (hunk 6 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: was just checking that both things where the same :)

File: src/main/scala/blast.scala (hunk 4 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: what exactly?

File: src/main/scala/blast.scala (hunk 3 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: yep

File: src/main/scala/blast.scala (hunk 2 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: it is not a blastn blastp whatever command. That's `allBlasts`

File: src/main/scala/blast.scala (hunk 1 above)
Comment (eparejatobes) [inline src/main/scala/blast.scala]: yes!

**Comment by eparejatobes** [comment]
Still missing options and fields, but I'm merging anyway. We can add them as we miss them.

**Pull request merged**